### ISSUE TITLE
StackOverflowException with .NET Core 2.2 in AddMvc()

### ISSUE THREAD

**Issue by ytzhakov** [seesharper/LightInject#462]
Just upgraded my ASP.NET Core MVC project from 2.1 to 2.2

LightInject throws StackOverflowException when calling app.UseMvc() or app.UseMvcWithDefaultRoute()
I was able to get rid of the issue by setting the Compatibility Version to 2.1.

Doesnt work:

services.AddMvc().AddControllersAsServices().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
app.UseMvcWithDefaultRoute(); //throws

Works:

services.AddMvc().AddControllersAsServices().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
app.UseMvcWithDefaultRoute(); //all good

Any ideas?

**Comment by seesharper** [comment]
Could you also post your startup.cs and program.cs?

**Comment by seesharper** [comment]
Unable to reproduce so far.

Here is my startup.cs 

```
public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {

            services.AddMvc().AddControllersAsServices().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
                app.UseHsts();
            }

            app.UseHttpsRedirection();
            app.UseMvcWithDefaultRoute();            
        }
    }
```

and my program.cs

```
 public class Program
    {
        public static void Main(string[] args)
        {
            CreateWebHostBuilder(args).Build().Run();
        }

        public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                .UseLightInject()
                .UseStartup<Startup>();
    }
```

**Comment by ytzhakov** [comment]
`public class Program
    {
        public static void Main(string[] args)
        {
            WebHost.CreateDefaultBuilder(args)
                   .UseStartup<Startup>()
                   .Build()
                   .Run();
        }
    }
`
  

`
public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        public IServiceProvider ConfigureServices(IServiceCollection services)
        {
            services.AddMvc()
                    .AddControllersAsServices()
                    .SetCompatibilityVersion(CompatibilityVersion.Version_2_2);

            services.AddCors(options =>
                {
                    options.AddPolicy("CorsPolicy",
                                      builder => builder.AllowAnyOrigin()
                                                        .AllowAnyMethod()
                                                        .AllowAnyHeader()
                                                        .AllowCredentials());
                });


            var containerOptions = new ContainerOptions { EnablePropertyInjection = false };
            var container = new ServiceContainer(containerOptions);
            
            IServiceProvider serviceProvider = container.CreateServiceProvider(services);

            return serviceProvider;
        }

        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDatabaseErrorPage();
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseHsts();
            }

            app.UseHttpsRedirection();
            app.UseStaticFiles();
            app.UseMvcWithDefaultRoute();
            app.UseCors("CorsPolicy");
        }
    }`

Maybe I should you .UseLightInject() in the program instead of what I use now from LightInject.Microsoft.DependencyInjection ?

**Comment by seesharper** [comment]
Yeah, that would be better indeed. You find that if you pull in the `LightInject.Microsoft.AspnetCore.Hosting` package

**Comment by ytzhakov** [comment]
![image](https://user-images.githubusercontent.com/1571884/49612277-e666fb80-f9ac-11e8-9c4f-dbb0d7a2280f.png)


UseLightInject() did not seem to replace the default service provider

**Comment by seesharper** [comment]
That is the container that is used to inject services into the startup class. 
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/startup?view=aspnetcore-2.2
If I remember correctly they operate with two containers. One for app services used during startup and one or the rest of the application.

**Comment by ytzhakov** [comment]
well I might be missing something really stupid but it doesn't seem to be the case.
I've created a new ASP.NET Core 2.2 project, used AddLightInject()
Then I requested IServiceProvider in ValuesController constructor and I'm still getting microsofts container.
Here's the code:

![image](https://user-images.githubusercontent.com/1571884/49645600-397c9500-fa25-11e8-94cc-e516c78e7351.png)

![image](https://user-images.githubusercontent.com/1571884/49645631-5913bd80-fa25-11e8-9f4b-ee245f1ecaef.png)


And here's what I get in ValuesController:

![image](https://user-images.githubusercontent.com/1571884/49645692-8ceee300-fa25-11e8-9fd9-7a0a4d8a6bc2.png)

**Comment by seesharper** [comment]
Hi again @ytzhakov and thanks for putting so much time into this.

There is a lot of more or less well documented conventions in aspnet core. 

It seems like you need this in your startup to make it replace the service provider 

```
 public void ConfigureContainer(IServiceContainer container)
        {
           
        }
``` 

But then we are back to a `StackoverflowException` when compatibility is set to 2_2.

Looking into it

**Comment by seesharper** [comment]
This is the service it stack overflows on 

```
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
Microsoft.Extensions.Options.IOptions`1[Microsoft.AspNetCore.Routing.EndpointOptions]
```

**Comment by seesharper** [comment]
Looks like this might be the class in question. 

Note: Just putting down my findings as I am debugging this 

https://github.com/aspnet/Extensions/blob/release/2.2/src/Options/Options/src/OptionsFactory.cs

**Comment by seesharper** [comment]
@ytzhakov I found it. It is related to `IEnumerable<T>` and variance.

Variance is on by default in LightInject. It means that `GetAllInstances<Foo>´ will bring back all instances of `Foo` and also `DerivedFoo`.  The problem here is  that some `IEnumerable<T>` somewhere deep in AspNetCore is resolved bringing back something that causes a StackOverFlowEx eption. You can turn this behaviour off by doing `containerOption.EnableVariance=false`. I need to talk to the AspNetCore team before I change the default here. I have a feeling it was not intentional to require this behaviour.

**Comment by ytzhakov** [comment]
Great.. it would really be interesting to find out whether or not this was intentional and if so, why... Please update this issue when you have more information

**Comment by AlexandrSitdikov** [comment]
I have a similar problem due to dependency injection.
Described my decision here https://github.com/seesharper/LightInject/issues/473
